public/app/core/components/RolePicker/RolePickerMenu.tsx

```diff
@@ -0,0 +1,585 @@
+                    onChange={onChange}
+                    // onClick={onMenuGroupClick}
+                    hideDescription
+                  />
+                ))}
+              </div>
+            </div>
+          )}
+          {!!customRoles.length && (
+            <div>
+              <div className={customStyles.groupHeader}>Custom roles</div>
+              <div className={styles.optionBody}>
+                {customRoles.map((option, i) => (
+                  <RoleMenuOption
+                    data={option}
+                    key={i}
+                    isSelected={
+                      !!(
+                        (option.uid && !!selectedOptions.find((opt) => opt.uid === option.uid)) ||
+                        !!selectedBuiltInRoles.find((role) => role.uid === option.uid)
+                      )
+                    }
+                    disabled={!!(option.uid && !!selectedBuiltInRoles.find((role) => role.uid === option.uid))}
+                    onChange={onChange}
+                    hideDescription
+                  />
+                ))}
+              </div>
+            </div>
+          )}
+        </CustomScrollbar>
+        <div className={customStyles.menuButtonRow}>
+          <HorizontalGroup justify="flex-end">
+            <Button size="sm" fill="text" onClick={onClearInternal}>
+              Clear all
+            </Button>
+            <Button size="sm" onClick={onUpdateInternal}>
+              Update
+            </Button>
+          </HorizontalGroup>
+        </div>
+      </div>
+      {showSubMenu ? (
+        <RolePickerSubMenu
+          options={subMenuOptions}
+          selectedOptions={selectedOptions}
+          disabledOptions={selectedBuiltInRoles}
+          onSelect={onChange}
+          onClear={onClearSubMenu}
+        />
+      ) : (
+        <div></div>
+      )}
+    </div>
+  );
+};
+
+const filterCustomRoles = (option: Role) => !option.name?.startsWith('fixed:');
+const filterFixedRoles = (option: Role) => option.name?.startsWith('fixed:');
+
+interface RolePickerSubMenuProps {
+  options: Role[];
+  selectedOptions: Role[];
+  disabledOptions: Role[];
+  onSelect: (option: Role) => void;
+  onClear?: () => void;
+}
+
+export const RolePickerSubMenu = ({
+  options,
+  selectedOptions,
+  disabledOptions,
+  onSelect,
+  onClear,
+}: RolePickerSubMenuProps): JSX.Element => {
+  const theme = useTheme2();
+  const styles = getSelectStyles(theme);
+  const customStyles = useStyles2(getStyles);
+
+  const onClearInternal = async () => {
+    if (onClear) {
+      onClear();
+    }
+  };
+
+  return (
+    <div className={customStyles.subMenu} aria-label="Role picker submenu">
+      <CustomScrollbar autoHide={false} autoHeightMax="300px" hideHorizontalTrack>
+        <div className={styles.optionBody}>
+          {options.map((option, i) => (
+            <RoleMenuOption
```

Comment:
Same rendering logic as for fixed and custom roles, so should be probably abstracted into a separate component.